# Use this notebook to time and subsequently decrease abm running time

In [1]:
import pandas as pd
import numpy as np
from ipywidgets import widgets, interact
from stockmarket import baselineCDoubleAuction, baselineCDA
import sqlite3
from tqdm import *

In [2]:
NRUNS = 1

# How much time does it take to run the model once

In [11]:
%time baselineCDoubleAuction.stockMarketSimulation(seed=NRUNS, simulation_time=100, amount_fundamentalists=40, amount_chartists=40, amount_firms=1,initial_money=(100, 200), initial_bid_ask=(1,1), initial_memory=(3,7),initial_ma_short=(1,3), initial_ma_long=(6,7), initial_profit=(200,200),initial_book_value=(10000, 10000), initial_stock_amount=400, order_expiration_time=120)

Before starting any improvements --> Wall time: 2min 22s

In [3]:
%time baselineCDA.stockMarketSimulation(seed=NRUNS, simulation_time=100, amount_fundamentalists=40, amount_chartists=40, amount_firms=1,initial_money=(100, 200), initial_bid_ask=(1,1), initial_memory=(3,7),initial_ma_short=(1,3), initial_ma_long=(6,7), initial_profit=(200,200),initial_book_value=(10000, 10000), initial_stock_amount=400, order_expiration_time=120)

Wall time: 461 ms


([trader_0,
  trader_1,
  trader_2,
  trader_3,
  trader_4,
  trader_5,
  trader_6,
  trader_7,
  trader_8,
  trader_9,
  trader_10,
  trader_11,
  trader_12,
  trader_13,
  trader_14,
  trader_15,
  trader_16,
  trader_17,
  trader_18,
  trader_19,
  trader_20,
  trader_21,
  trader_22,
  trader_23,
  trader_24,
  trader_25,
  trader_26,
  trader_27,
  trader_28,
  trader_29,
  trader_30,
  trader_31,
  trader_32,
  trader_33,
  trader_34,
  trader_35,
  trader_36,
  trader_37,
  trader_38,
  trader_39,
  trader_40,
  trader_41,
  trader_42,
  trader_43,
  trader_44,
  trader_45,
  trader_46,
  trader_47,
  trader_48,
  trader_49,
  trader_50,
  trader_51,
  trader_52,
  trader_53,
  trader_54,
  trader_55,
  trader_56,
  trader_57,
  trader_58,
  trader_59,
  trader_60,
  trader_61,
  trader_62,
  trader_63,
  trader_64,
  trader_65,
  trader_66,
  trader_67,
  trader_68,
  trader_69,
  trader_70,
  trader_71,
  trader_72,
  trader_73,
  trader_74,
  trader_75,
  trader_76,
  trader_

If I remove the recording of data in dataframes --> Wall time: 461 ms

# Then, we explore which functions take up the majority of time

In [12]:
%prun baselineCDoubleAuction.stockMarketSimulation(seed=NRUNS, simulation_time=100, amount_fundamentalists=40, amount_chartists=40, amount_firms=1,initial_money=(100, 200), initial_bid_ask=(1,1), initial_memory=(3,7),initial_ma_short=(1,3), initial_ma_long=(6,7), initial_profit=(200,200),initial_book_value=(10000, 10000), initial_stock_amount=400, order_expiration_time=120)

In [8]:

%timeit [emptyList.append(x) for x in range(10)]

100000 loops, best of 3: 1.88 µs per loop


In [38]:
def fillList(length):
    l = range(length)
    lst = []
    for x in l:
        lst.append(x)
    return lst

In [39]:
def fillArray(length):
    l = range(length)
    array = np.zeros(length)
    for x in l:
        array[x] = x
    return array

In [42]:
%time fillArray(1000000)

Wall time: 169 ms


array([  0.00000000e+00,   1.00000000e+00,   2.00000000e+00, ...,
         9.99997000e+05,   9.99998000e+05,   9.99999000e+05])

In [44]:
%time fillList(1000000)

In [40]:
%timeit fillArray(1000000)

10 loops, best of 3: 158 ms per loop


In [41]:
%timeit fillList(1000000)

10 loops, best of 3: 121 ms per loop


# Tips for optimization:

* multiple assignment is slower than individual assignment. For example "x,y=a,b" is slower than "x=a; y=b".
* chained comparisons are faster than using the "and" operator. Write "x < y < z" instead of "x < y and y < z
* Membership testing with sets and dictionaries is much faster, O(1), than searching sequences, O(n)
* String concatenation is best done with ''.join(seq) which is an O(n) process. 
* Function call overhead is large compared to other instructions.
* List comprehensions run a bit faster than equivalent for-loops (unless you're just going to throw away the result).

# Then, I try to remove the double use of . as much as possible. 

## In 